src/scales/scale.linearbase.js

```diff
@@ -18,13 +18,30 @@ function generateTicks(generationOptions, dataRange) {
 	var stepSize = generationOptions.stepSize;
 	var min = generationOptions.min;
 	var max = generationOptions.max;
-	var spacing, precision, factor, niceRange, niceMin, niceMax, numSpaces;
+	var spacing, precision, factor, niceMin, niceMax, numSpaces, maxNumSpaces;
 
 	if (stepSize && stepSize > 0) {
 		spacing = stepSize;
+		if (generationOptions.maxTicksLimit) {
+			maxNumSpaces = generationOptions.maxTicksLimit - 1;
+			// spacing is set to stepSize multiplied by a nice number of
+			// Math.ceil((max - min) / maxNumSpaces / stepSize) = num of steps that should be grouped
+			spacing *= helpers.niceNum(Math.ceil((dataRange.max - dataRange.min) / maxNumSpaces / stepSize));
+			numSpaces = Math.ceil(dataRange.max / spacing) - Math.floor(dataRange.min / spacing);
+			if (numSpaces > maxNumSpaces) {
+				// If the calculated num of spaces exceeds maxNumSpaces, recalculate it
+				spacing = helpers.niceNum(Math.ceil(numSpaces * spacing / maxNumSpaces / stepSize)) * stepSize;
+			}
+		}
 	} else {
-		niceRange = helpers.niceNum(dataRange.max - dataRange.min, false);
-		spacing = helpers.niceNum(niceRange / (generationOptions.maxTicks - 1), true);
+		maxNumSpaces = generationOptions.maxTicks - 1;
+		// spacing is set to a nice number of (max - min) / maxNumSpaces
+		spacing = helpers.niceNum((dataRange.max - dataRange.min) / maxNumSpaces);
+		numSpaces = Math.ceil(dataRange.max / spacing) - Math.floor(dataRange.min / spacing);
+		if (numSpaces > maxNumSpaces) {
+			// If the calculated num of spaces exceeds maxNumSpaces, recalculate it
+			spacing = helpers.niceNum(numSpaces * spacing / maxNumSpaces);
+		}
```

Comment:
These 2 blocks look very similar, I'm wondering if there is no way to simplify / refactor that logic. It's also weird that line 33 and 43 doesn't use the same rounding (ceil).

```javascript
var stepped = stepSize > 0;
var spacing = stepped ? stepSize : 1;
var maxNumSpaces = stepped ? maxTicksLimit : maxTicks;

// adjust spacing in case of maximum number of ticks
if (maxNumSpaces) {
    spacing *= //...;
    numSpaces = //...;
    if (numSpaces > maxNumSpaces) {
        spacing = //...;
    }
}

// compute factor from precision
if (precision) {
    //...
}
```

Reply:
The reason why line 33 and 43 are using different logics is that the former is looking for a nice number of how many steps should be grouped together into one tick interval while the latter is looking for a nice number of spacing between ticks. `Math.ceil` is used in the formar case as it must be an integer, but a value can be a floating point number in the latter case.

Reply:
I think the following simplification is possible as we know that `(max - min) / maxNumSpaces` is
- equal to or slightly less than `1` if `stepSize` is set and `maxTicksLimit` is not
- greater than `1` if both `stepSize` and `maxTicksLimit` is set

with the improvement in `getTickLimit` above, so we can eliminate `Math.ceil`.
```js
var maxNumSpaces = generationOptions.maxTicks - 1;
var unit = stepSize > 0 ? stepSize : 1;

spacing = helpers.niceNum((dataRange.max - dataRange.min) / maxNumSpaces / unit) * unit;
numSpaces = Math.ceil(dataRange.max / spacing) - Math.floor(dataRange.min / spacing);
if (numSpaces > maxNumSpaces) {
	spacing = helpers.niceNum(numSpaces * spacing / maxNumSpaces / unit) * unit;
}
```